### PR TITLE
Add registered derivatives to FParser hash

### LINKED ISSUE #1508: FParserAD::RegisterDerivative does not change function cache signature
`FParserAD::RegisterDerivative` is used by MOOSE to implement chain rule derivatives of material properties used in parsed expressions. Unfortunately if a material property is used first without declaring a nonlinear variable dependence (and thus not triggering a `FParserAD::RegisterDerivative` call) the derivatives of the function are cached on disk (in the `.jitcache` directory. If subsequently a nonlinear variable dependence for the material property is declared and `FParserAD::RegisterDerivative` is called on the function parser object the bytecode has is unaffected and the old derivative set is read from the persistent disk cache. This causes material property derivatives to be effectively ignored.

A workaround is to delete the `.jitcache` directory after making changes to the parsed function that relate to material property derivatives.

The fix I'm shooting for is to include the contents of the map build by `FParserAD::RegisterDerivative` in the hash signature of the function object.

See idaholab/moose#10333

### PR DESCRIPTION
**Pull request by dschwen**
This fixes #1508 (and thus idaholab/moose#10333). It avoids a hash collision that returns the wrong derivative after modifying a parsed function by adding a custom derivative mapping.

### CODE REVIEW

**Comment by dschwen** [comment]
Can I haz review?

File: tests/fparser/autodiff.C [hunk 1]
```diff
@@ -1,9 +1,9 @@
 #include "libmesh/fparser_ad.hh"
 
 // Ignore unused parameter warnings coming from cppunit headers
-#include <libmesh/ignore_warnings.h>
-#include <cppunit/extensions/HelperMacros.h>
 #include <cppunit/TestCase.h>
+#include <cppunit/extensions/HelperMacros.h>
+#include <libmesh/ignore_warnings.h>
```
Comment (roystgnr) [inline tests/fparser/autodiff.C]: Why'd the ignore_warnings header get moved?  Placement is important for that one...

**Comment by roystgnr** [comment]
:+1: other than the ignore_warnings weirdness

File: tests/fparser/autodiff.C (hunk 1 above)
Comment (dschwen) [reply tests/fparser/autodiff.C]: Ugh, I did edit my `.clang-format` (which had `DisableFormat: true` in it ) to include `SortIncludes: false`. But apparently only after saving that file while working on these edits.

File: tests/fparser/autodiff.C [hunk 2]
```diff
@@ -180,6 +182,48 @@ public:
...
+    dR.Optimize();
+    // dR = a + x*a
+
+    FunctionParserAD d2R(dR);
+    CPPUNIT_ASSERT_EQUAL (d2R.AutoDiff("x"), -1);
+    d2R.Optimize();
+    // d2R = 2*a + x*a
+
+    // we probe the parsers and check if they agree with the reference solution
+    for (x = -1.0; x < 1.0; x+=0.3726)
```
[the remark this answers is not in the document]
Comment (dschwen) [reply tests/fparser/autodiff.C]: There are carefully handcrafted artisanal loop increments made only from organic digits. You redneck wouldn't know anything about that.

[the remark this answers is not in the document]
Comment (jwpeterson) [reply tests/fparser/autodiff.C]: blinking_guy

[the remark this answers is not in the document]
Comment (roystgnr) [reply tests/fparser/autodiff.C]: To be fair, have you *seen* [the kind of crap that happens when we just rely on rand()](https://github.com/libMesh/libmesh/issues/1509) to do our pseudo-random testing?

[the remark this answers is not in the document]
Comment (dschwen) [reply tests/fparser/autodiff.C]: image

**Pull request merged**